### PR TITLE
Validate datetime against granules

### PR DESCRIPTION
**Pull request by binni979**
This is the pull request to solve [this](https://github.com/NASA-IMPACT/pyQuARC/issues/262) issue.

Overview: 
In the pyQuarc, while running the collection, it showed the warning message like this:
"The collection beginning date time `1983-07-01T00:00:00.000000Z` is not consistent with the first granule's beginning date time `1983-07-01T00:00:00.000Z`. Recommend updating the beginning date time to match the granule extent."

Changes made:
Modified the code to check and validate datetime to meet the same level of precision, now no warning shows up like the above. Tested on this 'C2068391958-LARC_ASDC' (umm-c)  collection to make sure the if the warning shows up.

### CODE REVIEW

File: pyQuARC/code/datetime_validator.py [hunk 1]
```diff
@@ -143,13 +143,39 @@ def validate_datetime_against_granules(
...
+                    continue
+            return None, False
+
+        # Compare the precision of the two datetime strings
+        # return get_precision(datetime_str1) == get_precision(datetime_str2)
         if len(granules["feed"]["entry"]) > 0:
             last_granule = granules["feed"]["entry"][0]
             last_granule_datetime = last_granule.get(time_key)
-            validity = datetime == last_granule_datetime
+            date_time, date_time_format = get_precision(datetime_string)
```
Comment (xhagrg) [inline pyQuARC/code/datetime_validator.py:174]: What is the use of `date_time_format`?

Comment (binni979) [reply pyQuARC/code/datetime_validator.py:174]: I return that date_time_format to see in which format date_time is converting, but in this code, it doesn't have any use so, I will remove it. Thank you for addressing that.

Comment (xhagrg) [reply pyQuARC/code/datetime_validator.py]: it is not updated?

Comment (binni979) [reply pyQuARC/code/datetime_validator.py]: Updated.

File: pyQuARC/code/datetime_validator.py [hunk 2]
```diff
@@ -143,13 +143,38 @@ def validate_datetime_against_granules(
...
+        # Define the formats in decreasing order of precision
+        formats = [
+            "%Y-%m-%dT%H:%M:%S.%f",  # Year to microsecond
+            "%Y-%m-%dT%H:%M:%S",  # Year to second
+            "%Y-%m-%dT%H:%M",  # Year to minute
+            "%Y-%m-%dT%H",  # Year to hour
+            "%Y-%m-%d",  # Year to day
+            "%Y-%m",  # Year to month
+            "%Y",  # Year
+        ]
```
Comment (xhagrg) [inline pyQuARC/code/datetime_validator.py:157]: Can move this to a constant/config

File: pyQuARC/code/datetime_validator.py [hunk 3]
```diff
@@ -143,13 +143,38 @@ def validate_datetime_against_granules(
...
+        # Function to determine the precision of a datetime string
+        def get_date_time(dt_str):
+            for fmt in formats:
+                try:
+                    date_time = datetime.strptime(dt_str, fmt)
+                    return date_time
+                except ValueError:
+                    continue
+            return None
+
```
Comment (xhagrg) [inline pyQuARC/code/datetime_validator.py:168]: Can be moved to utils?

File: pyQuARC/code/datetime_validator.py (hunk 2 above)
Comment (binni979) [reply pyQuARC/code/datetime_validator.py]: Yes, I can. I moved this to the constants.py file.

File: pyQuARC/code/datetime_validator.py (hunk 3 above)
Comment (binni979) [reply pyQuARC/code/datetime_validator.py]: Yes, I moved this to the utils file.

File: pyQuARC/code/utils.py [hunk 4]
```diff
@@ -64,3 +65,13 @@ def cmr_request(cmr_prms):
 
 def collection_in_cmr(cmr_prms):
     return cmr_request(cmr_prms).get("hits", 0) > 0
+
+
+def get_date_time(dt_str):
```
Comment (xhagrg) [inline pyQuARC/code/utils.py:70]: Please add doc strings as well.

Comment (binni979) [reply pyQuARC/code/utils.py:70]: Added doc strings as well.

File: pyQuARC/code/utils.py [hunk 5]
```diff
@@ -64,3 +65,20 @@ def cmr_request(cmr_prms):
...
+    parse using one of these formats will result in returning the corresponding `datetime` object.
+    If none of the formats match, the function returns `None`.
+    """
+    for fmt in DATE_FORMATS:
+        try:
+            date_time = datetime.strptime(dt_str, fmt)
+            return date_time
+        except ValueError:
+            continue
+    return None
```
Comment (xhagrg) [inline pyQuARC/code/utils.py:84]: This is not required.

**Review by xhagrg — APPROVED** [review]

**Review by xhagrg — APPROVED** [review]

**Pull request merged**